server/lib/scheduler/Schedule.ts

```diff
@@ -0,0 +1,142 @@
+import { map } from "asyncro";
+import * as _ from "lodash";
+import { UserSchedule } from "./UserSchedule";
+
+class Schedule {
+  private cacheForWhenTreeLastWorkedOn = {};
+  private rootParentCache;
+  private schedule = new Map;
+  private timeLimit;
+
+  public constructor({ rootParentCache, timeLimit }) {
+    this.rootParentCache = rootParentCache;
+    this.timeLimit = timeLimit;
+  }
+
+  private doesUserHaveASchedule(userId) {
+    return this.schedule.has(userId);
+  }
+
+  private createUserSchedule(userId) {
+    this.schedule.set(userId, new UserSchedule({ timeLimit: this.timeLimit, userId }));
+  }
+
+  private getUserSchedule(userId) {
+    return this.schedule.get(userId);
+  }
+
+  public async assignWorkspaceToUser(userId, workspaceId, startAtTimestamp = Date.now()) {
+    if (!this.doesUserHaveASchedule(userId)) {
+      this.createUserSchedule(userId);
+    }
+
+    const userSchedule = this.getUserSchedule(userId);
+    userSchedule.assignWorkspace(workspaceId, startAtTimestamp);
+
+    const rootParentId = await this.rootParentCache.getRootParentIdOfWorkspace(workspaceId);
+    this.cacheForWhenTreeLastWorkedOn[rootParentId] = startAtTimestamp;
+  }
+
+  public getMostRecentAssignmentForUser(userId) {
+    if (!this.doesUserHaveASchedule(userId)) {
+      return undefined;
+    }
+
+    const userSchedule = this.getUserSchedule(userId);
+    return userSchedule.getMostRecentAssignment();
+  }
+
+  /*
+    NOTE: this only considers the trees that at least one workspace in the
+    argument belong to
+  */
+  public async isInTreeWorkedOnLeastRecently(workspaceIds, workspaceId) {
+    const treesWorksOnLeastRecently = await this.getTreesWorkedOnLeastRecently(workspaceIds);
+    const rootParentId = await this.rootParentCache.getRootParentIdOfWorkspace(workspaceId);
+    return Boolean(treesWorksOnLeastRecently.find(rootWorkspaceId => rootWorkspaceId === rootParentId));
```

Comment:
Haven't seen Boolean used like this much. I'm used to using ```!!foo```, but I guess this should work the same, and is perhaps more straightforward.
That said, with lodash, you should have a fn to do just this.